tortoise/tortoise.go

```diff
@@ -1017,83 +812,70 @@ func (t *turtle) sumVotesForBlock(
 // consideration local opinion, it relies solely on global opinion.
 func (t *turtle) heal(ctx *tcontext, targetLayerID types.LayerID) {
 	// These are our starting values
-	pbaseOld := t.Verified
-	pbaseNew := t.Verified
+	pbaseOld := t.verified
+	pbaseNew := t.verified
 
 	for candidateLayerID := pbaseOld.Add(1); candidateLayerID.Before(targetLayerID); candidateLayerID = candidateLayerID.Add(1) {
 		logger := t.logger.WithContext(ctx).WithFields(
 			log.Named("old_verified_layer", pbaseOld),
 			log.Named("last_verified_layer", pbaseNew),
 			log.Named("target_layer", targetLayerID),
 			log.Named("candidate_layer", candidateLayerID),
-			log.Named("last_layer_received", t.Last),
-			log.Uint32("hdist", t.Hdist))
-
-		// we should never run on layers newer than Hdist back (from last layer received)
-		// when bootstrapping, don't attempt any verification at all
-		var latestLayerWeCanVerify types.LayerID
-		if t.Last.Before(types.NewLayerID(t.Hdist)) {
-			latestLayerWeCanVerify = mesh.GenesisLayer().Index()
-		} else {
-			latestLayerWeCanVerify = t.Last.Sub(t.Hdist)
-		}
-		if candidateLayerID.After(latestLayerWeCanVerify) {
-			logger.With().Error("cannot heal layer that's not at least hdist layers old",
-				log.Named("highest_healable_layer", latestLayerWeCanVerify))
-			return
-		}
+			log.Named("last_layer_received", t.last),
+			log.Uint32("hdist", t.Hdist),
+		)
 
 		// Calculate the global opinion on all blocks in the layer
 		// Note: we look at ALL blocks we've seen for the layer, not just those we've previously marked contextually valid
 		logger.Info("self healing attempting to verify candidate layer")
 
-		layerBlockIds, err := getLayerBlocksIDs(ctx, t.bdp, candidateLayerID)
-		if err != nil {
-			logger.Error("inconsistent state: can't find layer in database, cannot heal", log.Err(err))
-			return
-		}
+		// TODO(dshulyak) computation for expected weight should use last as an upper boundary
+		expected := computeExpectedWeight(t.epochWeight, candidateLayerID, t.processed)
+		threshold := weightFromUint64(0)
+		threshold = threshold.add(expected)
+		threshold = threshold.fraction(t.GlobalThreshold)
 
-		weight, err := computeExpectedVoteWeight(t.atxdb, t.epochWeight, candidateLayerID, t.Processed)
-		if err != nil {
-			logger.Error("failed to compute expected vote weight", log.Err(err))
-			return
-		}
-		logger.With().Info("healing: expected voting weight for layer", log.Stringer("weight", weight), candidateLayerID)
+		logger.With().Info("healing: expected voting weight at the layer",
+			log.Stringer("expected_weight", expected),
+			log.Stringer("threshold", threshold),
+			candidateLayerID,
+		)
 
 		// record the contextual validity for all blocks in this layer
-		for _, blockID := range layerBlockIds {
-			logger := logger.WithFields(log.Named("candidate_block_id", blockID))
+		for _, blockID := range t.blocks[candidateLayerID] {
```

Comment:
i am still not understanding why we don't reload all blocks from mesh

Reply:
why do we need to load them from mesh? we already loaded them in processLayer

it is a big mistake to have implicit inputs in the code that has a high bar for correctness. if we load more blocks here it means that the rest of the tortoise state won't know about them (e.g. blockLayer, votes and son on). so by seemingly innocent change, it introduces inconsistencies that make it hard to reason about tortoise correctness.

Reply:
by processing them explicitly i know that the state is consistent and there will no weird stuff when block is added in one place, but not added in another

Reply:
ok. i think i understand better now. whatever missing data in the normal run would be reloaded during rerun.